components/script/dom/bindings/structuredclone.rs

```diff
@@ -5,15 +5,131 @@
 //! This module implements structured cloning, as defined by [HTML]
 //! (https://html.spec.whatwg.org/multipage/#safe-passing-of-structured-data).
 
+use dom::bindings::conversions::root_from_handleobject;
 use dom::bindings::error::{Error, Fallible};
+use dom::bindings::js::Root;
+use dom::bindings::reflector::DomObject;
+use dom::blob::{Blob, BlobImpl};
 use dom::globalscope::GlobalScope;
-use js::jsapi::{HandleValue, MutableHandleValue};
-use js::jsapi::{JSContext, JS_ReadStructuredClone, JS_STRUCTURED_CLONE_VERSION};
-use js::jsapi::{JS_ClearPendingException, JS_WriteStructuredClone};
+use js::jsapi::{Handle, HandleObject, HandleValue, MutableHandleValue, JSAutoCompartment, JSContext};
+use js::jsapi::{JSStructuredCloneCallbacks, JSStructuredCloneReader, JSStructuredCloneWriter};
+use js::jsapi::{JS_ClearPendingException, JSObject, JS_ReadStructuredClone};
+use js::jsapi::{JS_ReadBytes, JS_WriteBytes};
+use js::jsapi::{JS_ReadUint32Pair, JS_WriteUint32Pair};
+use js::jsapi::{JS_STRUCTURED_CLONE_VERSION, JS_WriteStructuredClone};
+use js::jsapi::{MutableHandleObject, TransferableOwnership};
 use libc::size_t;
+use std::os::raw;
 use std::ptr;
 use std::slice;
 
+// TODO: Should we add Min and Max const to https://github.com/servo/rust-mozjs/blob/master/src/consts.rs?
+// TODO: Determine for sure which value Min and Max should have.
+// NOTE: Current values found at https://dxr.mozilla.org/mozilla-central/source/js/public/StructuredClone.h#323
+enum StructuredCloneTags {
+    /// To support additional types, add new tags with values incremented from the last one before Max.
+    Min = 0xFFFF8000,
+    DomBlob = 0xFFFF8001,
+    Max = 0xFFFFFFFF,
+}
+
+unsafe extern "C" fn read_blob(cx: *mut JSContext,
+                               r: *mut JSStructuredCloneReader,
+                               _data: u32)
+                               -> *mut JSObject {
+    let mut high: u32 = 0;
+    let mut low: u32 = 0;
+    assert!(JS_ReadUint32Pair(r, &mut high as *mut u32, &mut low as *mut u32));
+    let length = (low << high) as usize;
+    let mut buffer = vec![0u8; length];
+    assert!(JS_ReadBytes(r, buffer.as_mut_ptr() as *mut raw::c_void, length));
+    let target_global = GlobalScope::from_context(cx);
+    // NOTE: missing the content-type string here.
+    // Use JS_WriteString in write_callback? Make Blob.type_string pub?
```

Comment:
We should write and read the type string. We'll need a public method on Blob that exposes it, rather than making the member public.